File: zigzagConversion.py
import math
# Approach 1- String Traversal
def zigzagConversion(s,numRows): # TC O(n) // SC O(1)
	if numRows==1: return s 
	n=len(s)
	charsInSection=2*numRows-2
	ans=""
	for currRow in range(numRows):
		index=currRow

		while index<n:
			ans+=s[index]
			if currRow!=0 and currRow!=numRows-1:
				charsInBetween=charsInSection-2*currRow
				secondIndex=index+charsInBetween

				if secondIndex<n: 
					ans+=s[secondIndex]

			index+=charsInSection


	return ans 


# Approach 2- Simulate zigzag movement 
def zigzagConversion1(s,num_rows): # TC O(num_rows*n) // SC O(num_rows*n)
	if num_rows == 1:
		return s
        
	n = len(s)
	sections = math.ceil(n / (2 * num_rows - 2.0))
	num_cols = sections * (num_rows - 1)

	matrix = [[' '] * num_cols for _ in range(num_rows)]

	curr_row, curr_col = 0, 0
	curr_string_index = 0

	# Iterate in zig-zag pattern on matrix and fill it with string characters.
	while curr_string_index < n:
	    # Move down.
	    while curr_row < num_rows and curr_string_index < n:
	        matrix[curr_row][curr_col] = s[curr_string_index]
	        curr_row += 1
	        curr_string_index += 1
	        
	    curr_row -= 2
	    curr_col += 1
	    
	    # Move up (with moving right also).
	    while curr_row > 0 and curr_col < num_cols and curr_string_index < n:
	        matrix[curr_row][curr_col] = s[curr_string_index]
	        curr_row -= 1
	        curr_col += 1
	        curr_string_index += 1

	answer = ""
	for row in matrix:
	    answer += "".join(row)
	    
	return answer.replace(" ", "")

print(zigzagConversion("PAYPALISHIRING",3))
print(zigzagConversion1("PAYPALISHIRING",3))

File: zigzagConversion.java
import java.util.*;
// Approach 1- Simulate Zig-Zag movement
class Solution { // TC O(numRows*n) SC O(numRows*n)
    public String convert(String s, int numRows) {
        if(numRows==1) return s;
        int n=s.length();
        int sections= (int) Math.ceil(n/(2*numRows-2.0));
        int numCols=sections*(numRows-1);
        
        char[][] matrix=new char[numRows][numCols];
        for(char[] row: matrix){
            Arrays.fill(row, ' ');
        }
        
        int currRow=0, currCol=0, currentIndex=0;
        
        while(currentIndex<n){
            // down
            while(currRow<numRows && currentIndex<n){
                matrix[currRow++][currCol]=s.charAt(currentIndex++);
            }
            
            currRow-=2;
            currCol++;
            // diagonal
            while(currRow>0 && currCol<numCols && currentIndex<n){
                matrix[currRow--][currCol++]=s.charAt(currentIndex++);
            }
        }
        
        StringBuilder sb=new StringBuilder();
        for(char[] row: matrix){
            for(char c: row){
                if(c!=' '){
                    sb.append(c);
                }
            }
        }
        return sb.toString();
    }
}

// Approach 2- String Traversal
class Solution1{ // TC O(n) SC O(1)
    public String convert(String s, int numRows) {
        if(numRows==1) return s;
        
        StringBuilder sb=new StringBuilder();
        int n=s.length();
        int charsInSection=2*numRows-2;
        for(int currRow=0; currRow<numRows; currRow++){
            int index=currRow;
            while(index<n){
                sb.append(s.charAt(index));
                if(currRow!=0 && currRow!=numRows-1){
                    int charsInBetween=charsInSection-2*currRow;
                    int secondIndex=index+charsInBetween;
                    
                    if(secondIndex<n){
                        sb.append(s.charAt(secondIndex));
                    }
                }
                index+=charsInSection;
            }
        }
        return sb.toString();
    }
}

class zigzagConversion{
    public static void main(String[] args) {
        String x="HARSHILISHIRED";
        int rows=3;
        Solution s= new Solution();
        System.out.println(s.convert(x,rows));
        Solution1 s1= new Solution1();
        System.out.println(s1.convert(x,rows));
    }
}